tests/phpunit/PaymentMethods/WC_Stripe_UPE_Payment_Gateway_Test.php

```diff
@@ -3343,4 +3343,246 @@ public function test_add_bnpl_debug_metadata() {
 		$this->assertArrayHasKey( 'pmc_enabled', $result );
 		$this->assertEquals( 'yes', $result['pmc_enabled'] );
 	}
+
+	/**
+	 * Test that get_customer_id_for_order() correctly creates or updates customers with billing details.
+	 *
+	 * For guest users, billing details are retrieved from the order object.
+	 * For logged-in users, billing details come from user meta (user email and user meta fields),
+	 * with the order parameter available as a fallback when user data is missing.
+	 *
+	 * @dataProvider provide_get_customer_id_for_order_billing_details_test_cases
+	 *
+	 * @param string $scenario_name Description of the test scenario.
+	 * @param bool   $is_guest Whether the order is for a guest user.
+	 * @param string $existing_stripe_customer_id Existing Stripe customer ID for the user (empty for new customer).
+	 * @param string $expected_customer_id Expected Stripe customer ID to be returned.
+	 * @param string $api_url_pattern Pattern to match the API URL.
+	 * @param array  $billing_data Billing data to set on the order (and user meta for logged-in users).
+	 * @param array  $expected_customer_data Expected customer data in the API request.
+	 *
+	 * @return void
+	 */
+	public function test_get_customer_id_for_order_retrieves_billing_details_from_order( $scenario_name, $is_guest, $existing_stripe_customer_id, $expected_customer_id, $api_url_pattern, $billing_data, $expected_customer_data ) {
```

Comment:
Nit: why not specify explicit types?
```suggestion
	public function test_get_customer_id_for_order_retrieves_billing_details_from_order( string $scenario_name, bool $is_guest, string $existing_stripe_customer_id, string $expected_customer_id, string $api_url_pattern, array $billing_data, array $expected_customer_data ) {
```